### PR TITLE
SDI-301 Skeleton for new movements events

### PR DESCRIPTION
**Pull request by andymarke**
Will send a receive event for transfers

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/HmppsDomainEventEmitter.kt [hunk 1]
```diff
@@ -37,8 +39,7 @@ class HmppsDomainEventEmitter(
     differences: PrisonerDifferences,
   ) {
     runCatching {
-      PrisonerUpdatedEvent(offenderNo, differences.keys.toList().sorted())
-        .let { event -> PrisonerUpdatedDomainEvent(event, Instant.now(clock), diffProperties.host) }
+      PrisonerUpdatedDomainEvent(PrisonerUpdatedEvent(offenderNo, differences.keys.toList().sorted()), Instant.now(clock), diffProperties.host)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/HmppsDomainEventEmitter.kt:42]: IntelliJ warning asking to remove redundant `let`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/HmppsDomainEventEmitter.kt [hunk 2]
```diff
@@ -54,8 +55,7 @@ class HmppsDomainEventEmitter(
 
   fun emitPrisonerCreatedEvent(offenderNo: String) {
     runCatching {
-      PrisonerCreatedEvent(offenderNo)
-        .let { event -> PrisonerCreatedDomainEvent(event, Instant.now(clock), diffProperties.host) }
+      PrisonerCreatedDomainEvent(PrisonerCreatedEvent(offenderNo), Instant.now(clock), diffProperties.host)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/HmppsDomainEventEmitter.kt:58]: Same as above

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/HmppsDomainEventEmitter.kt [hunk 3]
```diff
@@ -69,10 +69,53 @@ class HmppsDomainEventEmitter(
...
+        .let { domainEvent ->
+          PublishRequest(topicArn, objectMapper.writeValueAsString(domainEvent))
+            .addMessageAttributesEntry("eventType", MessageAttributeValue().withDataType("String").withStringValue(PRISONER_RECEIVED_EVENT_TYPE))
+        }.also { publishRequest ->
+          topicSnsClient.publish(publishRequest)
+        }
+    }
+      .onFailure {
+        log.error("Failed to send event $CREATED_EVENT_TYPE for offenderNo=$offenderNo", it)
+        // TODO how do we retry without sending other messages?
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/HmppsDomainEventEmitter.kt:96]: will do a telemetry for manual recovery

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerDifferenceService.kt [hunk 4]
```diff
@@ -63,6 +64,7 @@ class PrisonerDifferenceService(
     if (prisonerHasChanged(offenderBooking.offenderNo, storedPrisoner)) {
       generateDiffEvent(existingPrisoner, offenderBooking, storedPrisoner)
       generateDiffTelemetry(existingPrisoner, offenderBooking, storedPrisoner)
+      prisonerMovementsEventService.generateAnyMovementEvents(existingPrisoner, storedPrisoner)
     }
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerDifferenceService.kt:68]: the meat; as well as generating the "update" event send movement events if required

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/HmppsDomainEventEmitter.kt [hunk 5]
```diff
@@ -120,3 +163,24 @@ data class PrisonerCreatedDomainEvent(
...
+) {
+  constructor(additionalInfo: PrisonerReceivedEvent, occurredAt: Instant, host: String) :
+    this(
+      additionalInfo = additionalInfo,
+      occurredAt = ISO_OFFSET_DATE_TIME.withZone(ZoneId.of("Europe/London")).format(occurredAt),
+      eventType = PRISONER_RECEIVED_EVENT_TYPE,
+      version = 1,
+      description = "A prisoner has been received into a prison",
+      detailUrl = ServletUriComponentsBuilder.fromUriString(host).path("/prisoner/{offenderNo}").buildAndExpand(additionalInfo.nomsNumber).toUri().toString(),
+    )
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/HmppsDomainEventEmitter.kt:185]: I'm guessing that we'll end up with a single `DomainEvent<E>` type once we have a few of these.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerMovementsEventService.kt [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import uk.gov.justice.digital.hmpps.prisonersearch.model.Prisoner
+import uk.gov.justice.digital.hmpps.prisonersearch.services.HmppsDomainEventEmitter
+import uk.gov.justice.digital.hmpps.prisonersearch.services.HmppsDomainEventEmitter.PrisonerReceiveReason.TRANSFERRED
+import uk.gov.justice.digital.hmpps.prisonersearch.services.diff.PossibleMovementChange.MovementChange.TransferIn
+
+@Service
+class PrisonerMovementsEventService(private val domainEventEmitter: HmppsDomainEventEmitter) {
+  fun generateAnyMovementEvents(
+    previousPrisonerSnapshot: Prisoner?,
+    prisoner: Prisoner
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerMovementsEventService.kt:13]: Nice improvement in language - be good to make it consistent across the old events too.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerDifferenceService.kt (hunk 4 above)
Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerDifferenceService.kt:68]: I guess this work will take a week or two - slightly worried that somebody will see the new events in Offender Events UI and start using them before they're ready. There are so many consumers from random managed services they're hard to keep track.

Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerDifferenceService.kt:68]: I'm not necessarily arguing for a feature switch btw - maybe just some BOLD text saying these events are WIP.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerMovementsEventServiceTest.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      this.firstName = "BOBBY"
+    }
+
+    prisonerMovementsEventService.generateAnyMovementEvents(previousPrisonerSnapshot, prisoner)
+
+    verifyNoInteractions(domainEventsEmitter)
+  }
+
+  @Test
+  internal fun `will emit anything for a new prisoner `() {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerMovementsEventServiceTest.kt:36]: Is this missing a "not"?

**Review by mikehalmamoj — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/HmppsDomainEventEmitter.kt (hunk 5 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/HmppsDomainEventEmitter.kt:185]: 100%

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerMovementsEventServiceTest.kt (hunk 7 above)
Comment (andymarke) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerMovementsEventServiceTest.kt:36]: yes!

**Comment by andymarke** [comment]
@mikehalmamoj I'll merge it and just double check at least the transfer event is being emitted.

I think this might only take a couple of days but lets see

**Pull request merged**